### PR TITLE
Fix size inflation in spa_get_worst_case_asize()

### PR DESCRIPTION
**Pull request by loli10K**
### Description

While i was working to improve ashift handling in https://github.com/zfsonlinux/zfs/pull/5763 i encountered a strange issue: it seems that with higher ashift values we can easily push enough writes to trigger a failure (ENOSPC) in the ZIO pipeline: this will suspend all the I/O on the pool because we can't allocate more space on the storage (`zio_dva_allocate()->metaslab_alloc()->metaslab_alloc_dva()`).

This happens on both 0.6.5.X and 0.7.0-rcX tagged releases and also on HEAD, even after merging _OpenZFS 7793 - ztest fails assertion in dmu_tx_willuse_space_ (3ec3bc2).

The codepath that should prevent us from writing more data than we should be allowed to is:

```
zfs_write()
 + dmu_tx_assign()
    + dmu_tx_try_assign()
       + dsl_dir_tempreserve_space()
          + dsl_dir_tempreserve_impl()
```

`dmu_tx_try_assign()` relies on `spa_get_worst_case_asize()` to inflate our writes size before handing it off to `dsl_dir_tempreserve_space()` which will check if we're allowed to allocate more space or should get back an ENOSPC. Since, with the current implementation, `spa_get_worst_case_asize()` just inflates our writes by a factor of `spa_asize_inflation` (24) we can easily write more data than we should be allowed to (in open context) when these conditions are met:

1. pool has vdevs with ashift >= 13
2. dataset we're writing to has recordsize=512
3. dataset we're writing to has copies=3

Now, every write that gets committed to the DMU will grow from 512 bytes (recordsize) to 8K (ashift=13) * 3 (copies) = 24K, which is a factor of 48. Remembering `spa_asize_inflation` is just 24 we just wrote double the amount of data we should be allowed to.

This can be easily reproduced on both Linux:

```
root@debian-8-zfs:~# cat /sys/module/zfs/version 
0.7.0-rc3_155_g08dc05f
root@debian-8-zfs:~# 
root@debian-8-zfs:~# POOLNAME='testpool'
root@debian-8-zfs:~# TMPDIR='/var/tmp'
root@debian-8-zfs:~# #
root@debian-8-zfs:~# fallocate -l 512m $TMPDIR/zpool.dat
root@debian-8-zfs:~# zpool create -f $POOLNAME $TMPDIR/zpool.dat -o ashift=13 -o failmode=panic
root@debian-8-zfs:~# zfs create $POOLNAME/fs
root@debian-8-zfs:~# zfs set recordsize=512 $POOLNAME/fs
root@debian-8-zfs:~# zfs set compression=off $POOLNAME/fs
root@debian-8-zfs:~# zfs set copies=3 $POOLNAME/fs
root@debian-8-zfs:~# #
root@debian-8-zfs:~# sysctl -w kernel.panic=1
kernel.panic = 1
root@debian-8-zfs:~# #
root@debian-8-zfs:~# # suspend the pool
root@debian-8-zfs:~# time dd if=/dev/urandom of=/$POOLNAME/fs/zero bs=1024M count=1
[  634.857293] Kernel panic - not syncing: Pool 'testpool' has encountered an uncorrectable I/O failure and the failure mode property for this pool is set to panic.
[  634.859043] CPU: 0 PID: 1956 Comm: z_null_int Tainted: P           OE   4.6.4 #5
[  634.859866] Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
[  634.860527]  0000000000000086 000000001a5122b1 ffffffff8146a545 ffffffffc0418ec8
[  634.861410]  ffff880204d47d50 ffffffff81171b2f ffff880200000010 ffff880204d47d60
[  634.862570]  ffff880204d47cf8 000000001a5122b1 0000000000000001 ffff880208d5e000
[  634.864119] Call Trace:
[  634.864510]  [<ffffffff8146a545>] ? dump_stack+0x5c/0x77
[  634.865083]  [<ffffffff81171b2f>] ? panic+0xdf/0x226
[  634.865666]  [<ffffffffc033095f>] ? zio_suspend+0x25f/0x260 [zfs]
[  634.866394]  [<ffffffffc0332e49>] ? zio_done+0x15c9/0x18f0 [zfs]
[  634.867086]  [<ffffffffc0327fd0>] ? zio_execute+0x100/0x320 [zfs]
[  634.867761]  [<ffffffffc014c5f3>] ? taskq_thread+0x293/0x580 [spl]
[  634.868429]  [<ffffffff810a37a0>] ? wake_up_q+0x60/0x60
[  634.868932]  [<ffffffffc014c360>] ? taskq_thread_spawn+0x50/0x50 [spl]
[  634.869542]  [<ffffffff81098ddf>] ? kthread+0xdf/0x100
[  634.870061]  [<ffffffff81802eb2>] ? ret_from_fork+0x22/0x40
[  634.870647]  [<ffffffff81098d00>] ? kthread_park+0x50/0x50
[  634.871327] Dumping ftrace buffer:
[  634.871694]    (ftrace buffer empty)
[  634.872022] Kernel Offset: disabled
[  634.872683] Rebooting in 1 seconds..
```

and Illumos:

```
- SmartOS (build: 20170315T210557Z)
[root@52-54-00-d3-7a-01 ~]# cd /var/crash/volatile/
[root@52-54-00-d3-7a-01 /var/crash/volatile]# savecore -f vmdump.0 
savecore: System dump time: Tue Mar 28 16:44:13 2017

savecore: saving system crash dump in /var/crash/volatile/{unix,vmcore}.0
Constructing namelist /var/crash/volatile/unix.0
Constructing corefile /var/crash/volatile/vmcore.0
 0:34 100% done: 104142 of 104142 pages saved
[root@52-54-00-d3-7a-01 /var/crash/volatile]# mdb 0
Loading modules: [ unix genunix specfs dtrace mac cpu.generic uppc apix scsi_vhci ufs ip hook neti sockfs arp usba uhci stmf_sbd stmf zfs mm lofs idm sata random cpc logindmux ptm sppp nfs sd ]
> $C
ffffff0008283960 vpanic()
ffffff00082839a0 zio_suspend+0xa9(ffffff024e3cd000, ffffff024f90f420)
ffffff0008283a40 zio_done+0x635(ffffff024f90f420)
ffffff0008283a70 zio_execute+0x7f(ffffff024f90f420)
ffffff0008283b30 taskq_thread+0x2d0(ffffff02500d56e0)
ffffff0008283b40 thread_start+8()
> ::zio_state -r
ADDRESS                 TYPE  STAGE            WAITER           TIME_ELAPSED
ffffff024f90f420        NULL  FAILED           ffffff000b1f6c40 -            
 ffffff024fee23e8       WRITE FAILED           -                -            
  ffffff0250d84408      WRITE DONE             -                -            
   ffffff024fe110a0     WRITE DONE             -                -            
    ffffff024fc7b090    WRITE DONE             -                -            
     ffffff024f9af008   WRITE DONE             -                -            
      ffffff024fee8c28  WRITE FAILED           -                -            
       ffffff024fe5bbe0 WRITE FAILED           -                -            
       ffffff024fa3a7f0 WRITE FAILED           -                -            
[...]
> ffffff024fee8c28::print zio_t io_error
io_error = 0x1c
> ffffff024fee8c28::print zio_t io_spa | ::print spa_t spa_min_ashift
spa_min_ashift = 0xd
> ffffff024fee8c28::print zio_t io_spa | ::print spa_t spa_max_ashift
spa_max_ashift = 0xd
> ffffff000b1f6c40::whatis
ffffff000b1f6c40 is allocated as a thread structure
> ffffff000b1f6c40::stacks -v
mdb: stacks: processing kernel threads
mdb: stacks: 1 unique stacks / 1 threads
mdb: stacks: done
THREAD           STATE    SOBJ                COUNT
ffffff000b1f6c40 SLEEP    CV                      1
                 swtch+0x141
                 cv_wait+0x70
                 zio_wait+0x6b
                 dsl_pool_sync+0xb3
                 spa_sync+0x386
                 txg_sync_thread+0x227
                 thread_start+8

> 
```

If all this is true i think we can fix this by multiplying `spa_asize_inflation` and `MAX(lsize, 1<<spa->spa_max_ashift)` in `spa_get_worst_case_asize()`.

Side remark: the ZIO pipeline is really "intricate" to troubleshoot, i literally lost sleep hours over this issue.

Also "Requires-builders: none" because i need someone to validate all this before we pass it to the buildbot pipeline.

### Motivation and Context

I need this change so we can safely bump ashift max value.

### How Has This Been Tested?

Hours of pain and suffering with gdb/systemtap and mdb/dtrace. Will add a test to the ZTS if needed.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: module/zfs/spa_misc.c [hunk 1]
```diff
@@ -1617,7 +1617,7 @@ spa_freeze_txg(spa_t *spa)
 uint64_t
 spa_get_worst_case_asize(spa_t *spa, uint64_t lsize)
 {
-	return (lsize * spa_asize_inflation);
+	return (MAX(lsize, 1<<spa->spa_max_ashift) * spa_asize_inflation);
```
Comment (behlendorf) [inline module/zfs/spa_misc.c]: [nit] add a space before and after the `<<` operator.

**Comment by loli10K** [comment]
@behlendorf the test case will not work on Illumos though, we need to be able to `zpool create -o ashift=x` to reproduce this.

**Comment by behlendorf** [comment]
@loli10K that's unfortunate, but it will still be useful to have the test case under Linux until this added functionality get's adopted in OpenZFS.

**Comment by loli10K** [comment]
I went to add a new "enospc_003_pos" test case and found "enospc_002_pos" wasn't included in our ZTS runfile.

With this commit "enospc_002_pos" fails when we try to `zfs create testpool/fs` on a full pool:

```
[  412.520717] VERIFY3(0 == dmu_tx_assign(tx, TXG_WAIT)) failed (0 == 28)
[  412.521382] PANIC at dsl_synctask.c:83:dsl_sync_task()
[  412.521949] Showing stack for process 3259
[  412.521952] CPU: 0 PID: 3259 Comm: zfs Tainted: P           OE   4.6.4 #5
[  412.521953] Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
[  412.521954]  0000000000000000 00000000afcee0ad ffffffff8146a545 ffffffffc0439b48
[  412.521957]  ffff880210f6fcb8 ffffffffc014a9df ffff88002eaec030 ffff880200000030
[  412.521960]  ffff880210f6fcc8 ffff880210f6fc68 2833594649524556 756d64203d3d2030
[  412.521962] Call Trace:
[  412.521968]  [<ffffffff8146a545>] ? dump_stack+0x5c/0x77
[  412.521976]  [<ffffffffc014a9df>] ? spl_panic+0xbf/0xf0 [spl]
[  412.522017]  [<ffffffffc02ac100>] ? dmu_objset_clone_sync+0x240/0x240 [zfs]
[  412.522020]  [<ffffffff8113e4e9>] ? stack_trace_call+0x49/0x80
[  412.522022]  [<ffffffff818009b0>] ? __mutex_unlock_slowpath+0xd0/0x180
[  412.522024]  [<ffffffffc0014067>] ? 0xffffffffc0014067
[  412.522053]  [<ffffffffc0314c89>] ? txg_rele_to_sync+0x89/0x110 [zfs]
[  412.522076]  [<ffffffffc02ac100>] ? dmu_objset_clone_sync+0x240/0x240 [zfs]
[  412.522121]  [<ffffffffc02ac100>] ? dmu_objset_clone_sync+0x240/0x240 [zfs]
[  412.522147]  [<ffffffffc02e5aa5>] ? dsl_sync_task+0x1d5/0x250 [zfs]
[  412.522169]  [<ffffffffc02b0270>] ? dmu_objset_create_impl+0x330/0x330 [zfs]
[  412.522201]  [<ffffffffc034aa60>] ? zfs_ioc_hold+0x120/0x120 [zfs]
[  412.522223]  [<ffffffffc02abe34>] ? dmu_objset_create+0x64/0x80 [zfs]
[  412.522254]  [<ffffffffc034aa60>] ? zfs_ioc_hold+0x120/0x120 [zfs]
[  412.522285]  [<ffffffffc034fef3>] ? zfs_ioc_create+0x143/0x2e0 [zfs]
[  412.522316]  [<ffffffffc0350723>] ? zfsdev_ioctl+0x2c3/0x740 [zfs]
[  412.522320]  [<ffffffff81206b79>] ? do_vfs_ioctl+0x99/0x5d0
[  412.522322]  [<ffffffff81206ae5>] ? do_vfs_ioctl+0x5/0x5d0
[  412.522324]  [<ffffffff81207126>] ? SyS_ioctl+0x76/0x90
[  412.522326]  [<ffffffff81802c76>] ? system_call_fast_compare_end+0xc/0x96
```

which is:

```c
    66	dsl_sync_task(const char *pool, dsl_checkfunc_t *checkfunc,
    67	    dsl_syncfunc_t *syncfunc, void *arg,
    68	    int blocks_modified, zfs_space_check_t space_check)
    69	{
    70		spa_t *spa;
    71		dmu_tx_t *tx;
    72		int err;
    73		dsl_sync_task_t dst = { { { NULL } } };
    74		dsl_pool_t *dp;
    75	
    76		err = spa_open(pool, &spa, FTAG);
    77		if (err != 0)
    78			return (err);
    79		dp = spa_get_dsl(spa);
    80	
    81	top:
    82		tx = dmu_tx_create_dd(dp->dp_mos_dir);
    83		VERIFY0(dmu_tx_assign(tx, TXG_WAIT));
```

triggered by `log_mustnot_expect space $ZFS create $TESTPOOL/$TESTFS/subfs`: se we **do** expect to fail, but this should be handled more gracefully.

Needs more investigation.

**Comment by behlendorf** [comment]
@loli10K the `redundancy_001_pos` test case appears to be failing for a similar reason.

File: module/zfs/spa_misc.c [hunk 2]
```diff
@@ -1614,11 +1614,20 @@ spa_freeze_txg(spa_t *spa)
...
+ * pool we use its value instead, since the write will end up using the whole
+ * block anyway.
+ */
 uint64_t
 spa_get_worst_case_asize(spa_t *spa, uint64_t lsize)
 {
-	return (lsize * spa_asize_inflation);
+	if (lsize == 0) {
+		return (0);	/* No inflation needed */
+	}
```
Comment (behlendorf) [inline module/zfs/spa_misc.c]: nit: please drop the `{ }` around the single line conditional.

**Comment by behlendorf** [comment]
@loli10K can you please rebase this on master and add your signed-off-by.  OpenZFS 7290 has been merged so you'll also need to update your test case accordingly ($ZFS -> zfs, etc).

After that it looks good to merge, unless @ahrens has any remaining concerns.

**Comment by loli10K** [comment]
@behlendorf the "Signed-off-by:" is already there though, i don't know why the BB is failing on that check.

**Pull request merged**